README.md
# Functional Interpeter
An Interpreter for a Simple Functional Language

<img src="https://cdn-images-1.medium.com/max/2600/1*ntgBBkoEipqEHD-NpgAspg.png">

You would write a `compose` function (.) as follows.
```Lambda "compose" (Lambda "g" (Lambda "x" (FunCall (Variable "f") (FunCall (Variable "g") (Variable "x))))```

Define a function which adds one to argument ((+1))
```(Lambda "plusOne" (Add (Variable "x") (Number 1)))```

Compose our plusOnes twice which evaluates to a function which adds 2 to the argument.
```(FunCall (FunCall (Variable "compose") (Variable "plusOne")) (Variable "plusOne"))```


ChangeLog.md
# Changelog for functional-interpreter

## Unreleased changes


test/Spec.hs
import Test.Hspec

import qualified Data.Map.Lazy as M
import Data.Map.Lazy (Map)

import Control.Monad.Except
import Control.Monad.State

import Data.Either
import Prelude

import Interpreter

main :: IO ()
main = hspec step1

step1 =
  describe "Step 1" $
  it "Eval can read bound variables" $
  runExcept (evalStateT (eval expr) Scope) `shouldBe` Right (NumVal 65)
  where
    Scope = Scope $ M.fromList [("x", 42), ("y", 23)]
    expr = Add (Variable "x") (Variable "y")

step2 =
  describe "Step 2" $
  it "Eval can bind variables" $
  runExcept (evalStateT (eval expr) emptyScope) `shouldBe` Right (NumVal 65)
  where
    expr = Let "x" (Number 42) (Add (Variable "x") (Number 23))

step3 =
  describe "Step3" $
  it "Can apply lambdas" $
  runExcept (evalStateT (eval exp) emptyScope) `shouldBe` Right (NumVal 42)
  where
    exp =
      (Let
         "f"
         (Lambda "x" (Add (Variable "x") (Variable "x")))
         (FunCall (Variable "f") (Number 21)) -- 42
       )


src/A.hs
module A where

import Control.Monad
import Control.Monad.Except
import Control.Monad.State

import Data.Either
import qualified Data.Map.Lazy as M
import Data.Map.Lazy (Map)
import Data.Maybe

import Prelude

-- Step 1:
data Exp
  = Number Int
  | Variable String
  | Add Exp
        Exp
  | Mul Exp
        Exp
  deriving (Show)

type Err = String

newtype Memory =
  Memory (Map String Int)
  deriving (Show)

data VarBinding =
  VarBinding String
             Int
  deriving (Show)

eval :: Exp -> StateT Memory (Except Err) Int
eval exp@(Add a b) = add a b
eval exp@(Mul a b) = mul a b

emptyMemory :: Memory
emptyMemory = Memory M.empty

lookupVarBinding :: String -> StateT Memory (Except Err) Int
lookupVarBinding name = do
  (Memory mem) <- get
  case M.lookup name mem of
    (Just varBinding) -> return varBinding
    Nothing -> throwError ("Unbound Variable: " ++ name)
  where
    mem = M.fromList [("x", 42), ("y", 23)]

getBindingVal :: VarBinding -> Int
getBindingVal (VarBinding _ val) = val

--lookupVarVal :: String -> StateT Memory (Except Err) (Either Err Int)
--lookupVarVal name = return $ (gets (lookupVarBinding name))
binaryOp :: (Int -> Int -> Int) -> Exp -> Exp -> StateT Memory (Except Err) Int
binaryOp op (Variable nameA) (Variable nameB) = do
  a <- lookupVarBinding nameA
  b <- lookupVarBinding nameB
  return $ a `op` b
binaryOp op (Variable nameA) (Number b) = do
  a <- lookupVarBinding nameA
  return $ a `op` b
binaryOp op (Number a) (Variable nameB) = do
  b <- lookupVarBinding nameB
  return $ a `op` b
binaryOp op (Number a) (Number b) = return $ a `op` b

add :: Exp -> Exp -> StateT Memory (Except Err) Int
add = binaryOp (+)

mul :: Exp -> Exp -> StateT Memory (Except Err) Int
mul = binaryOp (*)
-- test.hs
-- import TomsCode
-- let res = eval (Map.ofList [("x", 42), ("y", 23)]) (Add (Variable "x") (Variable "y")) -- 65
-- assert_eq res 65


src/Interpreter.hs
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE DeriveFunctor #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE StandaloneDeriving #-}

module Interpreter where

import Control.Monad
import Control.Monad.Except
import Control.Monad.Reader

import Data.Either
import Data.Map.Lazy (Map)
import qualified Data.Map.Lazy as M
import Data.Maybe

import Prelude

data Exp a where
  Number :: Int -> Exp Int
  Boolean :: Bool -> Exp Bool
  Variable :: String -> Exp a
  Add :: Exp Int -> Exp Int -> Exp Int
  Mul :: Exp Int -> Exp Int -> Exp Int
  IfThenElse :: Exp Bool -> Exp a -> Exp a -> Exp a
  Or :: Exp Bool -> Exp Bool -> Exp Bool
  And :: Exp Bool -> Exp Bool -> Exp Bool
  Not :: Exp Bool -> Exp Bool
  Eq :: Exp a -> Exp b -> Exp Bool
  Let :: String -> Exp a -> Exp b -> Exp c -- let name = exp in exp
  FunCall :: Exp String -> Exp a -> Exp b -- not tyypesafe !!!
  Lambda :: String -> Exp a -> Exp b -- not typesafe !!!

deriving instance Show (Exp a)

--
--newtype EvalM a = EvalM
--  { runEvalM :: Scope -> Either LangErr a
--  } deriving (Functor, Applicative, Monad, MonadReader Scope)
data Value
  = NumVal Int
  | BoolVal Bool
  | FunVal (Value -> ReaderT Scope (Except LangErr) Value)

instance Show Value where
  show (FunVal _) = "FunVal String -> Value"
  show (NumVal n) = show n

unNumVal :: Value -> Int
unNumVal (NumVal n) = n
unNumVal x = error (show x ++ " is not a NumVal")

unBoolVal :: Value -> Bool
unBoolVal (BoolVal n) = n
unBoolVal x = error (show x ++ " is not a BoolVal")

data LangErr =
  LangErr Err
          (Maybe Value)

instance Show LangErr where
  show (LangErr err val) = show err ++ " in the expression: \n" ++ show val

data Err
  = UnboundVar String
  | TypeError TypeErr
  | VarAlreadyBound String
  deriving (Show)

data TypeErr
  = ExpectedFunction
  | ExpectedNumVal
  | ExpectedBoolVal
  deriving (Show)

newtype Scope =
  Scope (Map String Value)
  deriving (Show)

emptyScope :: Scope
emptyScope = Scope M.empty

eval :: Exp a -> ReaderT Scope (Except LangErr) Value
eval (Number n) = return $ NumVal n
eval (Variable name) = lookupVarBinding name
eval (Boolean b) = return $ BoolVal b
eval (And a b) = binaryOp and' a b
eval (Or a b) = binaryOp or' a b
eval (Not a) = do
  valA <- eval a
  either throwError return (not' valA)
eval (IfThenElse cond a b) = do
  condVal <- eval cond
  ifThenElse condVal a b
eval (Add a b) = binaryOp add a b
eval (Mul a b) = binaryOp mul a b
eval (Let varName varExp body) = do
  varVal <- eval varExp
  bindVar varName varVal (eval body)
eval (Lambda paramName bodyExp) = defineLambda paramName bodyExp
eval (FunCall funcName arg) = do
  func <- eval funcName
  apply func arg

defineLambda :: String -> Exp a -> ReaderT Scope (Except LangErr) Value
defineLambda paramName body = do
  currScope <- ask
  return $ FunVal $ \val -> bindVar paramName val (eval body)

apply :: Value -> Exp a -> ReaderT Scope (Except LangErr) Value
apply (FunVal func) arg = eval arg >>= \a -> func a
apply val _ = throwError $ LangErr (TypeError ExpectedFunction) (pure val)

bindVar ::
     String
  -> Value
  -> ReaderT Scope (Except LangErr) Value
  -> ReaderT Scope (Except LangErr) Value
bindVar name val = local (\(Scope m) -> Scope $ M.insert name val m)

binaryOp ::
     (Value -> Value -> Either LangErr Value)
  -> Exp a
  -> Exp a
  -> ReaderT Scope (Except LangErr) Value
binaryOp op a b = do
  valA <- eval a
  valB <- eval b
  either throwError return (op valA valB)

lookupVarBinding :: String -> ReaderT Scope (Except LangErr) Value
lookupVarBinding name = do
  Scope mem <- ask
  case M.lookup name mem of
    Just varBinding -> return varBinding
    Nothing -> throwError $ LangErr (UnboundVar name) Nothing

add :: Value -> Value -> Either LangErr Value
add (NumVal a) (NumVal b) = Right $ NumVal $ a + b
add a (NumVal _) = Left $ LangErr (TypeError ExpectedNumVal) (pure a)
add (NumVal _) b = Left $ LangErr (TypeError ExpectedNumVal) (pure b)
add _ _ = Left $ LangErr (TypeError ExpectedNumVal) Nothing

mul :: Value -> Value -> Either LangErr Value
mul (NumVal a) (NumVal b) = Right $ NumVal $ a * b
mul a (NumVal _) = Left $ LangErr (TypeError ExpectedNumVal) (pure a)
mul (NumVal _) b = Left $ LangErr (TypeError ExpectedNumVal) (pure b)
mul _ _ = Left $ LangErr (TypeError ExpectedNumVal) Nothing

and' :: Value -> Value -> Either LangErr Value
and' (BoolVal a) (BoolVal b) = Right $ BoolVal $ a && b
and' _ _ = Left $ LangErr (TypeError ExpectedBoolVal) Nothing

or' :: Value -> Value -> Either LangErr Value
or' (BoolVal a) (BoolVal b) = Right $ BoolVal $ a || b
or' _ _ = Left $ LangErr (TypeError ExpectedBoolVal) Nothing

not' :: Value -> Either LangErr Value
not' (BoolVal a) = Right $ BoolVal $ not a
not' _ = Left $ LangErr (TypeError ExpectedBoolVal) Nothing

eq' :: Value -> Value -> Either LangErr Value
eq' (BoolVal a) (BoolVal b) = Right $ BoolVal $ a == b
eq' _ _ = Left $ LangErr (TypeError ExpectedBoolVal) Nothing

primFunc ::
     (Value -> Either LangErr Value)
  -> Exp a
  -> ReaderT Scope (Except LangErr) Value
primFunc f expA = do
  valA <- eval expA
  either throwError return (f valA)

primFunc2 ::
     (Value -> Value -> Either LangErr Value)
  -> Exp a
  -> Exp a
  -> ReaderT Scope (Except LangErr) Value
primFunc2 f expA expB = do
  valA <- eval expA
  primFunc (f valA) expB

neg :: Exp Int -> ReaderT Scope (Except LangErr) Value
neg =
  primFunc
    (\case
       NumVal i -> Right $ NumVal $ negate i
       a -> Left $ LangErr (TypeError ExpectedNumVal) (Just a))

compose :: Value -> Value -> ReaderT Scope (Except LangErr) Value
compose =
  \case
    a@(NumVal _) -> error $ show $ LangErr (TypeError ExpectedFunction) Nothing
    FunVal f ->
      \case
        b@(NumVal _) ->
          error $ show $ LangErr (TypeError ExpectedFunction) Nothing
        FunVal g -> return $ FunVal (f <=< g)

ifThenElse :: Value -> Exp a -> Exp a -> ReaderT Scope (Except LangErr) Value
ifThenElse (BoolVal bool') expA expB =
  if bool'
    then eval expA
    else eval expB
ifThenElse e _ _ = throwError (LangErr (TypeError ExpectedBoolVal) (pure e))

typeOf :: Value -> String
typeOf (BoolVal _) = "BoolVal"
typeOf (NumVal _) = "NumVal"
typeOf _ = "FunVal"


ex-eval.hs
-- Step 1:

data Exp
    = Number Integer
    | Variable String
    | Add Exp Exp
    | Mul Exp Exp

eval :: Map String Integer -> Exp -> Integer


-- test.hs
-- import TomsCode
-- let res = eval (Map.ofList [("x", 42), ("y", 23)]) (Add (Variable "x") (Variable "y")) -- 65
-- assert_eq res 65

-- Step 2:

data Exp
    = Number Integer
    | Variable String
    | Add Exp Exp
    | Mul Exp Exp
    | Let String Exp Exp -- let name = exp in exp


eval {} (Let "x" 42 (Add (Variable x) (Number 23))) -- 65


-- Step 3:

data Exp
    = Number Integer
    | Variable String
    | Add Exp Exp
    | Mul Exp Exp
    | Let String Exp Exp -- let name = exp in exp
    | Funcall Exp [Exp]
    | Lambda String Exp

data Value = NumVal Integer | FunVal (String -> Value)

eval :: Map String Value -> Value

eval {} (Let "f" (Lambda "x" (Add (Variable x) (Variable x))) (FunCall (Variable "f") [Number 21])) -- 42

-- Not part of the exercise
data Type
    = IntegerType
    | FunctionType Type Type
    | TypeVariable String

typeOf :: Expr -> Maybe Type

typeOf (Add (Number 1) (Number 2)) -- Some IntegerType
typeOf (Add (Number 1) (Lambda "x" (Variable "x"))) -- None because type error
typeOf (Lambda "x" (Add (Integer 1) (Variable "x"))) -- FunctionType IntegerType IntegerType

typeOf (Lambda "x" (Variable "x")) -- FunctionType (TypeVariable "a") (TypeVariable "a")